vnext/Microsoft.ReactNative/Views/Impl/ScrollViewViewChanger.cpp

```diff
@@ -0,0 +1,98 @@
+  }
+
+  if (m_horizontal) {
+    scrollViewer.ChangeView(scrollViewer.ScrollableWidth(), nullptr, nullptr, !animated);
+  } else {
+    scrollViewer.ChangeView(nullptr, scrollViewer.ScrollableHeight(), nullptr, !animated);
+  }
+}
+
+void ScrollViewViewChanger::OnViewChanging(
+    const xaml::Controls::ScrollViewer &scrollViewer,
+    const xaml::Controls::ScrollViewerViewChangingEventArgs &args) {
+  // For non-inverted views, a ScrollViewer.ViewChanging event always emits an `onScroll` event
+  if (m_inverted) {
+    // For inverted views, we need to detect if we're scrolling to or away from the bottom edge to enable or disable
+    // view anchoring
+    const auto scrollingToEnd =
+        IsScrollingToEnd(scrollViewer, args.NextView().HorizontalOffset(), args.NextView().VerticalOffset());
+
+    // Do not update scroll anchoring while scrolling to end
+    if (!m_isScrollingToEnd) {
+      ScrollViewUWPImplementation(scrollViewer).ContentAnchoringEnabled(!scrollingToEnd);
+      if (m_wasViewAnchoringEnabled && scrollingToEnd) {
+        // If scrolling to anchor edge, turn off view anchoring
+        SetContentScrollAnchors(scrollViewer, false);
+        m_wasViewAnchoringEnabled = false;
+      } else if (!m_wasViewAnchoringEnabled && !scrollingToEnd) {
+        // If scrolling away from anchor edge, turn on view anchoring
+        SetContentScrollAnchors(scrollViewer, true);
+        m_wasViewAnchoringEnabled = true;
+      }
+    }
+  }
+}
+
+void ScrollViewViewChanger::OnViewChanged(
+    const xaml::Controls::ScrollViewer &scrollViewer,
+    const xaml::Controls::ScrollViewerViewChangedEventArgs &args) {
+  // Stop tracking scroll-to-end once the ScrollView comes to rest
+  if (!args.IsIntermediate() && m_isScrollingToEnd) {
+    m_isScrollingToEnd = false;
+    if (!IsScrollingToEnd(scrollViewer, scrollViewer.HorizontalOffset(), scrollViewer.VerticalOffset())) {
+      ScrollViewUWPImplementation(scrollViewer).ContentAnchoringEnabled(true);
+      SetContentScrollAnchors(scrollViewer, true);
+    }
+  }
+}
+
+void ScrollViewViewChanger::SetContentScrollAnchors(const xaml::Controls::ScrollViewer &scrollViewer, bool enabled) {
+  const auto snapPointManager = scrollViewer.Content().as<SnapPointManagingContentControl>();
+  auto panel = snapPointManager->Content().as<xaml::Controls::Panel>();
+  for (auto child : panel.Children()) {
+    const auto childElement = child.as<xaml::UIElement>();
+    if (enabled) {
+      childElement.CanBeScrollAnchor(true);
+    } else {
+      childElement.ClearValue(xaml::UIElement::CanBeScrollAnchorProperty());
+    }
+  }
+}
+
+bool ScrollViewViewChanger::IsScrollingToEnd(
+    const xaml::Controls::ScrollViewer &scrollViewer,
+    double horizontalOffset,
+    double verticalOffset) {
+  return m_horizontal ? horizontalOffset > (scrollViewer.ScrollableWidth() - SCROLL_EPSILON)
+                      : verticalOffset > (scrollViewer.ScrollableHeight() - SCROLL_EPSILON);
+}
```

Comment:
Trying to better understand this logic, and why we do not anchor when scrolling to the end. Is it to special case allowing new content to show up?

Reply:
If we call the React ScrollView.scrollToEnd command, we presumably don't want anything to interrupt the scroll operation (except perhaps the user, but that's non-trivial to detect). This worked around a specific bug we saw where calling `scrollToEnd` on the ScrollView would land at the second item in the list, rather than the first item in the list when the first item arrived while scrolling. Typically with VirtualizedList, when a new item is mounted at the head of the list data, one item will be unmounted from the tail of the list data, and I think that causes a view port shift when anchoring is enabled.

Reply:
And actually, it wasn't just for the imperative `ScrollView.scrollToEnd` command, I believe bottom edge anchoring was broken unless we disabled content anchoring as well.

Reply:
Could the same bug happen for offset based scroll to index?

Seems like we kind of conceptually want a batch of "scrolling to target" to "target reached and list settled" where we do not want to anchor.